### PR TITLE
Faster tests

### PR DESCRIPTION
**Pull request by PAUL-BERNARD**
This PR targets [3 of the slowest tests](https://github.com/deepinv/deepinv/actions/runs/31991013911/job/95274487263#step:8:999).
Time comparisons are done on my machine, on CPU and GPU.

#### `test_sampling::test_sde`
Before: 2300s
After: 360s
`test_diffusion_reproducibility`: 27s

- Parametrization to avoid big nested loops
- Reduction of number of steps (10 -> 2) since we don't test performances
- Moved seed reproducibility test to a new test: `test_diffusion_reproducibility`
- Added a `device` argument to `FourierEmbedding` since I got an error about tensors being on different devices

#### `test_models::test_denoiser_sigma_{gray, color}`

Before: 630s for gray, 1140s for color
After: 97s for gray, 190s for color

- Only test with `batch_size=1 and 2` (not 3)
- Image size from 64x64 -> 16x16
Note: this test is slow especially because of DEAL. I'm opening a separate PR for this.

#### `test_optim::test_least_squares_implicit_backward`

Before: 5400s
After: 8s

- Removed call to `torch.gradcheck`
- Compare gradient of `least_squares_implicit_backward` and `least_squares`
- Use of a manual seed (this was a flaky tests that would regularly fail the CI, this seed seems to work on my machine). See #1168 

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` and `ruff check .` run successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [-] Updated docstrings related to the changes (as applicable).
- [-] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### LLM policy

LLM usage is ok, but not PRs generated 100% by AI. See our [LLM policy](https://deepinv.org/contributing.html#llm-policy) Tick below as appropriate:

- [x] I did not use LLM tools to write the code
- [ ] I, a human, wrote code, and LLM tools helped me to write part of the code.
- [ ] An LLM tool wrote all of the code.
- [ ] An agent submitted the PR and wrote the description.

### CODE REVIEW

File: deepinv/models/ncsnpp.py [hunk 1]
```diff
@@ -152,7 +152,7 @@ def __init__(
         self.map_noise = (
             PositionalEmbedding(num_channels=noise_channels, endpoint=True)
             if embedding_type == "positional"
-            else FourierEmbedding(num_channels=noise_channels)
+            else FourierEmbedding(num_channels=noise_channels, device=device)
```
Comment (romainvo) [inline deepinv/models/ncsnpp.py:155]: Can you also pass `device=device` for the other modules that support it, e.g. `Linear` ?

File: deepinv/tests/test_models.py [hunk 2]
```diff
@@ -531,9 +531,9 @@ def test_denoiser_1_channel(imsize_1_channel, device, denoiser):
 
 
 @pytest.mark.parametrize("denoiser", MODEL_LIST_1_CHANNEL)
-@pytest.mark.parametrize("batch_size", [1, 2, 3])
```
Comment (romainvo) [inline deepinv/tests/test_models.py:534]: I would keep batch size at least [1,2] if there are potentially different behaviours?

File: deepinv/tests/test_models.py [hunk 3]
```diff
@@ -560,9 +560,9 @@ def test_denoiser_sigma_gray(batch_size, denoiser, device):
 
 
 @pytest.mark.parametrize("denoiser", MODEL_LIST)
-@pytest.mark.parametrize("batch_size", [1, 2, 3])
```
Comment (romainvo) [inline deepinv/tests/test_models.py:563]: Same remark above

File: deepinv/tests/test_sampling.py [hunk 4]
```diff
@@ -274,130 +291,90 @@ def test_build_algo(algo, imsize, device):
...
+        **kwargs,
+    )
+    assert sample.shape == (2, 3, 64, 64)
+
+    # Test posterior sampling
+    posterior = PosteriorDiffusion(
+        data_fidelity=DPSDataFidelity(denoiser=denoiser),
+        sde=sde,
+        denoiser=NCSNpp(device=device),
+        solver=EulerSolver(timesteps=timesteps, rng=rng),
```
Comment (romainvo) [inline deepinv/tests/test_sampling.py]: Not sure why we initially run instantiate only with `denosiers[0]` and `solvers[0]` but I would instantiate `PosteriorDiffusion` with the current `solver` and `denoiser` if it works


```suggestion
        denoiser=denoiser
        solver=solver,
```

File: deepinv/tests/test_sampling.py [hunk 5]
```diff
@@ -274,130 +291,90 @@ def test_build_algo(algo, imsize, device):
...
+        physics,
+        x_init=(2, 3, 64, 64),
+        seed=111,
+    )
+    # Test output shape
+    assert x_hat.shape == (2, 3, 64, 64)
+
+    del denoiser
+    gc.collect()
+    torch.cuda.empty_cache()
```
Comment (romainvo) [inline deepinv/tests/test_sampling.py]: Maybe add a little comment explaining that pytest does not clean objects properly between runs (or is the cause something else?)

**Review by romainvo — CHANGES_REQUESTED** [review]
Thanks @PAUL-BERNARD for tackling this! Simple curiosity, by how much is it reducing the tests execution? My remarks are in the comments

File: deepinv/tests/test_sampling.py [hunk 6]
```diff
@@ -274,130 +291,90 @@ def test_build_algo(algo, imsize, device):
...
-                    get_trajectory=True,
-                    **kwargs,
-                )
-                x_init_2 = trajectory[0]
-                # Test reproducibility
-                assert torch.allclose(x_init_1, x_init_2, atol=1e-5, rtol=1e-5)
-                assert (
-                    torch.nn.functional.mse_loss(sample_1, sample_2, reduction="mean")
-                    < 1e-2
-                )
```
Comment (romainvo) [inline deepinv/tests/test_sampling.py:361]: Maybe we can do an isolated reproducibility test, where we run it with only 1 solver and 1 denoiser?

File: deepinv/tests/test_sampling.py (hunk 4 above)
Comment (PAUL-BERNARD) [reply deepinv/tests/test_sampling.py]: I just tried, but with denoiser=ADMUNet, OOM errors are coming back (with 8GB VRAM), I'm looking for a solution

File: deepinv/tests/test_sampling.py (hunk 6 above)
Comment (PAUL-BERNARD) [reply deepinv/tests/test_sampling.py:361]: Done :)

File: deepinv/tests/test_sampling.py (hunk 5 above)
Comment (PAUL-BERNARD) [reply deepinv/tests/test_sampling.py]: Done

File: deepinv/tests/test_sampling.py (hunk 4 above)
Comment (PAUL-BERNARD) [reply deepinv/tests/test_sampling.py]: Done, this was a bit more work than expected, since we need to pass kwargs for ADMUNet

**Comment by PAUL-BERNARD** [comment]
> Simple curiosity, by how much is it reducing the tests execution?

## On my computer (cumulated time on GPU and CPU):

#### `test_sampling::test_sde`
Before: 2300s
After: 360s
`test_diffusion_reproducibility`: 27s

#### `test_models::test_denoiser_sigma_{gray, color}`

Before: 630s for gray, 1140s for color
After: 97s for gray, 190s for color

#### `test_optim::test_least_squares_implicit_backward`

Before: 5400s
After: 8s

## On the CI

#### `test_optim::test_least_squares_implicit_backward`

5400s -> 10s

#### `test_sampling::test_sde`

1040s -> 300s

#### `test_models::test_denoiser_sigma_gray/color`

480s -> 530s ( weird, it should be lower, it should be lower with the DEAL PR )

File: deepinv/tests/test_sampling.py [hunk 7]
```diff
@@ -274,130 +291,154 @@ def test_build_algo(algo, imsize, device):
...
+            seed=111,
+            **kwargs,
+        )
+        # Test output shape
+        assert x_hat.shape == (2, 3, 64, 64)
+    finally:
+        # pytest seems to not clean objects properly, which can cause OOM errors.
+        del denoiser, sde, posterior
+        gc.collect()
+        torch.cuda.empty_cache()
```
Comment (Tmodrzyk) [inline deepinv/tests/test_sampling.py:379]: Hm this seems odd no ? Could it be related to the SDE / diffusion implementation instead ? 
Or is this documented somewhere for pytest ? 
It might be worth to investigate this in a separate issue / PR. :)

File: deepinv/tests/test_sampling.py [hunk 8]
```diff
@@ -274,130 +291,154 @@ def test_build_algo(algo, imsize, device):
...
+    # Test output shape
+    assert x_hat_1.shape == (2, 3, 64, 64)
+    # Test reproducibility
+    x_hat_2 = posterior(
+        y,
+        physics,
+        x_init=(2, 3, 64, 64),
+        seed=111,
+    )
+    assert torch.nn.functional.mse_loss(x_hat_1, x_hat_2, reduction="mean") < 1e-2
```
Comment (Tmodrzyk) [inline deepinv/tests/test_sampling.py]: I believe we use `torch.allclose` in other tests, could we use that here too just for consistency (departing from the original implem) ?

**Review by Tmodrzyk — COMMENTED** [review]
Hi @PAUL-BERNARD thank you for these incredible speed-ups !
I'm eager to see the positive impact on the CI. 
I have only a few minor comments, otherwise LGTM.

File: deepinv/tests/test_sampling.py (hunk 7 above)
Comment (PAUL-BERNARD) [reply deepinv/tests/test_sampling.py:379]: It is, I can't find any documentation / issue for this, and I don't know if this comes pytorch or pytest, but I can see the memory increasing with every new test run up to the OOM error.
It's worth investigating but I have spent enough time on this for this PR and this seems to fix the issue

File: deepinv/tests/test_sampling.py (hunk 8 above)
Comment (PAUL-BERNARD) [reply deepinv/tests/test_sampling.py]: Done (with `torch.testing.assert_close` instead of `assert torch.allclose` which is equivalent but with a better error message)

File: deepinv/tests/test_sampling.py (hunk 7 above)
Comment (Tmodrzyk) [reply deepinv/tests/test_sampling.py:379]: Of course, this is non-blocking to merge this PR.
Could you fill an issue so that we keep a trace of this odd memory leak ? :smile:

**Review by romainvo — APPROVED** [review]
LGTM, thanks @PAUL-BERNARD
